[1단계 - 로또 구현] 포포(이민우) 미션 제출합니다.

Minuring opened the pull request:
안녕하세요 7기 BE 포포입니다.
학습로그 : https://prolog.techcourse.co.kr/studylogs/3838

# 신경쓴 점
1. WinningLotto가 프로그램에서 유일하게 존재하는 당첨 로또라는 의미로 싱글톤 패턴으로 구현하였는데, 우선 테스트할 때 어려움을 겪었습니다.
2. Getter와 Setter를 최대한 지양하기 위해 Lotto와 WinningLotto 사이에서 양방향 메서드를 사용하거나 DTO를 도입하였습니다. 구현하고보니 getter는 사용하지 않게 되었지만 양방향 메서드를 통한 Lotto와 WinningLotto가 서로를 참조하게 되는 상황입니다. 이에 대해서 그냥 getter를 일부 허용하는 것이 더 나은 방법일지 혹은 다른 개선 방안이 있을 지 궁금합니다.

jinu0328 committed:
docs(readme): 1차 요구사항 업데이트

jinu0328 committed:
feat(view): 사용자 입력 처리하는 InputView 구현

- 구입 금액 입력 받는 함수 getPurchaseInput 구현

jinu0328 committed:
feat(view): 출력 담당 OutputView 구현

- 구매개수 출력 메서드 구현

jinu0328 committed:
feat(model): Lotto 모델 생성

- 기본 검증 메서드 구현(개수, 범위, 중복)
- 당첨번호 일치, 보너스번호 일치 판단 메서드 구현

jinu0328 committed:
feat(util): 난수 생성 유틸 클래스 구현

jinu0328 committed:
feat(service): 로또 생성 담당 LottoGenerateService 구현

- 생성된 난수 오름차순 정렬하여 Lotto 객체 생성
- 구입 금액 유효성 검사 메서드 구현

jinu0328 committed:
feat(model): Lotto 일급 컬렉션 Lottos 구현

jinu0328 committed:
feat(dto): LottoDto 구현

- Lotto 클래스 내에 toDto 메서드 구현

jinu0328 committed:
feat(view): 로또 번호 출력 메서드 구현

jinu0328 committed:
feat(util): 사용자 입력을 유효값으로 변환하는 InputConverter 클래스 구현

jinu0328 committed:
feat(view): 당첨번호 입력 & 보너스번호 입력 메서드 구현

jinu0328 committed:
feat(model): WinningLotto 모델 클래스 구현

- 싱글턴 패턴 적용
- 보너스 번호 중복 및 범위 유효성 검사 메서드 구현

jinu0328 committed:
feat(model): PrizeTier enum 구현

jinu0328 committed:
feat(model): 당첨 통계를 나타내는 Statistics 모델 구현

jinu0328 committed:
feat(model): Lotto의 PrizeTier 산출 메서드 구현

- Lotto와 WinningLotto 객체의 양방향 메서드로 구현

jinu0328 committed:
feat(service): StatisticsService 구현

jinu0328 committed:
feat(model): PrizeTier 별 개수 & 총 상금 계산 메서드 구현

jinu0328 committed:
feat(dto): 로또 출력을 위한 LottoDto 구현

jinu0328 committed:
feat(dto): Statistics 출력을 위한 StatisticsDto 구현

jinu0328 committed:
feat(model): PrizeTier getter 추가

jinu0328 committed:
chore(model): Statistics의 사용되지 않는 멤버 제거

jinu0328 committed:
feat(view): 당첨 통계 출력 메서드 구현

jinu0328 committed:
docs(readme): 1차 요구사항 구현 완료

jinu0328 committed:
feat(factory): Lotto 생성 담당 클래스 LottoFactory 구현

jinu0328 committed:
feat(factory): Lottos 생성 담당 클래스 LottosFactory 구현

jinu0328 committed:
feat(dto): LottosDto 구현

jinu0328 committed:
feat(service): 로또 추첨 담당 서비스 LottoEvaluationService 구현

jinu0328 committed:
feat(view): ViewFacade 구현

jinu0328 committed:
feat(config): view 및 service 객체 의존성 관리 담당 Container 구현

jinu0328 committed:
feat(controller): LottoController 구현

- 분리된 메서든 간 데이터 전달을 위해 record를 사용

jinu0328 committed:
feat(app): 프로그램 시작점 Application 구현

jinu0328 committed:
chore(view): 출력에 줄바꿈 추가

jinu0328 committed:
fix(view): scanner.next -> scanner.nextLine

jinu0328 committed:
fix(util): 로또 번호 정렬 방식 수정

- toList()를 통한 List는 불변이므로 사전에 정렬 처리

jinu0328 committed:
chore(util): 공백 제거

jinu0328 committed:
fix(service): 구입 금액이 0일 시 오류 방지

jinu0328 committed:
feat(model): 로또 등수 결정 메서드 & dto 변환 메서드 구현

jinu0328 committed:
chore(model): Lotto 모델의 사용되지 않는 멤버 제거

jinu0328 committed:
chore(model): PrizeTier의 사용되지 않는 메서드 제거

jinu0328 committed:
chore(model): WinningLotto의 멤버 이름 변경

- lotto -> basicLotto

jinu0328 committed:
refactor(model): 등수에 해당하는 로또 개수 로직 변경

- 가독성을 위해 for, if문에서 스트림으로 변경

jinu0328 committed:
feat(factory): Statistics 생성 담당 StatisticsFactory 구현

jinu0328 committed:
refactor: 가독성을 위한 메서드 분리

jinu0328 committed:
refactor(error): 에러메시지 Enum 구현 및 적용

jinu0328 committed:
test(model): Lotto 단위 테스트 구현

jinu0328 committed:
test(model): PrizeTier 단위 테스트 구현

jinu0328 committed:
test(model): WinningLotto 단위테스트 구현

Minuring committed:
docs(readme): README 수정

Hyunta commented on README.md:
```diff
@@ -1,3 +1,55 @@
+
+- [ ]  발행된 로또의 개수를 산출하여 출력한다
+- [ ]  발행된 로또의 번호를 출력한다
+- [ ]  당첨 통계를 출력한다
```
요구사항을 꼼꼼하게 잘 작성해주셨네요!

Minuring replied:
감사합니다!

Hyunta commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,52 @@
+
+    public boolean isBonusMatched(int bonusNumber) {
+        return numbers.contains(bonusNumber);
+    }
```
기능을 개발한 커밋과 테스트를 만든 커밋이 분리되어 있네요.
개발을 하고 테스트를 만들지 않고, 테스트를 일괄적으로 만드신 이유가 있을까요?

Minuring replied:
테스트 또한 기능 개발의 일부로 생각은 했는데 테스트 코드 작성법이 미숙해서, 마감기한을 지키려다 보니 일괄적으로 만들게 되었습니다!
앞으로는 기능 개발과 테스트를 하나의 단위로 만드려고 합니다!

Hyunta commented on README.md:
```diff
@@ -55,4 +55,4 @@
 - [x]  발행된 로또의 번호를 출력한다
-- [x]  당첨 통계를 출력한다
\ No newline at end of file
+- [x]  당첨 통계를 출력한다
```
구현 완료 커밋은 필요없지 않을까요?
커밋 로그를 작성하는 의미에 대해서 한번 생각해보면 좋을 것 같습니다.
동료 개발자가 어떻게 커밋을 해주면 서로에게 도움이 될까요?

Minuring replied:
> 구현 완료 커밋은 필요없지 않을까요?

질문을 보고 생각을 많이 해 봤는데, 잘 이해가 되지 않습니다 😭
구현을 완료했다는 커밋 로그를 남길 필요가 없다는 의미인가요?
아니면, 기능 목록(Readme)에 구현 완료 여부를 기록할 필요가 없다는 의미인가요?

---
커밋 로그를 통해 변경 이력들을 보고 무엇이 어떤 이유로 변경이 되어왔는 지 알기 위해 커밋 로그를 작성하는 것 같다고 생각합니다!
어떤 내용을 어떤 이유로 변경하게 되었는 지를 알려주는 커밋을 하면 서로에게 도움이 될 것 같습니다!

Hyunta replied:
![image](https://github.com/user-attachments/assets/5e42253d-7044-4d65-877e-95ac3d1b1696)

사진을 하나 첨부해드릴걸 그랬네요ㅎㅎ
구현 부분과 완료 처리한 부분의 커밋 로그가 분리되어져 있어서 남긴 커멘트였습니다.

Minuring replied:
아하 이해했습니다! 한 개의 커밋으로 올리겠습니다!

Hyunta commented on src/main/java/factory/StatisticsFactory.java:
```diff
@@ -0,0 +1,11 @@
+package factory;
```
패키지 분류가 꽤 많네요.
factory는 왜 따로 분류됐는지 알 수 있을까요?

Minuring replied:
도메인 객체들에 대한 생성 책임을 팩토리에게 맡기고자 한 의도였습니다.
지금 다시 보니 오버엔지니어링이라고 생각됩니다!
지금처럼 객체 생성 방법이 단순한 상황이면 팩토리를 이용하는 의미가 없는 것 같습니다.

만약 예를들어 숫자로만 이루어지지 않은 복권 등 여러 종류의 `Lotto`가 있었고 각각의 생성 방법이 달랐다면 팩토리 분류가 의미를 가졌을 것이라고 생각됩니다.
이 경우 팩토리를 통해 호출자는 재료(파라미터)만 넘기고 구체적인 생성 방법을 알 필요 없이 재료에 맞는 `Lotto`를 생성할 수 있기 때문입니다.

Minuring replied:
패키지 분류 관점으로만 본다면 2번이 더 나은 방법일까요?
1번의 경우 패키지명을 보고 프로젝트 구조를 읽기 쉬울 것 같고,
2번의 경우 객체 생성을 팩토리가 담당한다고 했을 때 같은 패키지에 있으므로 protected로 기본 생성자를 통한 객체 생성을 막을 수 있다는 점인 것 같습니다.

1. 현재
```bash
├── factory
│   └── LottoFactory.java
└── model
    └── Lotto.java
```

2. 모델과 함께 두기
```bash
└── model
    └── lotto
        ├── Lotto.java
        └── LottoFactory.java
```

Hyunta replied:
경우에 따라 달라질 것 같습니다.
만약 도메인이 100개 있는데 팩토리가 1개 있다면 2번처럼 했을 것 같아요.
반대로 팩토리가 100개 있고 도메인이 1개있다면 1번처럼 했을 것 같습니다.

패키지 분류에 정답은 없습니다.
같이 일하는 사람이 더 쉽게 클래스를 찾을 수 있고 불편함을 느끼지 않도록 구성하려고 시도해보시면 될 것 같아요.

Hyunta commented on src/test/java/model/LottoTest.java:
```diff
@@ -0,0 +1,66 @@
+    void 범위를_벗어난_번호가_오면_예외가_발생한다(List<Integer> numbers) {
+        assertThrows(IllegalArgumentException.class, () -> new Lotto(numbers));
+    }
+}
```
파일 끝에 개행이 안되어있어서 아래 그림처럼 나오네요.
한번 학습해보시면 좋을 것 같습니다.

![image](https://github.com/user-attachments/assets/b677b147-355a-4ad7-938f-3f3341ee048b)

Minuring replied:
미처 생각하지 못 한 부분이네요!
덕분에 파일 마지막에 개행이 있어야 하는 이유도 알았습니다 수정하겠습니다 👍

Hyunta commented on src/test/java/model/WinningLottoTest.java:
```diff
@@ -0,0 +1,46 @@
+        );
+    }
+
+    @Order(1)
+    @ParameterizedTest
+    @MethodSource("provideBasicLottoAndInvalidBonusNumber")
+    void 유효하지_않은_보너스번호가_들어오면_예외를_발생시킨다(Lotto basicLotto, int bonusNumber) {
+        assertThrows(IllegalArgumentException.class, () -> WinningLotto.initialize(basicLotto, bonusNumber));
+    }
+
+    @Order(2)
+    @ParameterizedTest
+    @MethodSource("provideBasicLottoAndBonusNumber")
+    void 정상적인_보너스번호가_들어오면_예외를_발생시키지않는다(Lotto basicLotto, int bonusNumber) {
+        assertDoesNotThrow(() -> WinningLotto.initialize(basicLotto, bonusNumber));
+    }
```
테스트가 순서에 의존하도록 나왔다는 것은 설계가 잘못됐을 가능성이 높아보입니다.
아직 코드 구조에 대한 리뷰를 드리지는 않았는데 찾아서 수정해볼 수 있을까요?
해결하기 어려우시면 의심이 되는 부분을 공유해주시면 같이 탐구해봅시다.

Minuring replied:
WinningLotto를 싱글톤으로 구현함에 따라 테스트가 불가능했습니다.
```
한 번 초기화된 당첨 로또는 다시 초기화할 수 없다는 제약을 의도적으로 추가하였고,
WinningLotto 클래스는 매 테스트마다가 아닌 처음 한 번만 로드되므로,
한 번 초기화되고 나면 초기화 메서드를 실행 시 `IllegalStateException`이 발생해 테스트가 실패했습니다.
```

그래서 마감시간 이슈로 사실상 해결을 피하고자 순서를 지정했습니다..! 바로 고치겠습니다

Hyunta commented on README.md:
```diff
@@ -1,3 +1,58 @@
 # java-lotto
 
 로또 미션 저장소
+
```
# 질문에 대한 답변

> WinningLotto가 프로그램에서 유일하게 존재하는 당첨 로또라는 의미로 싱글톤 패턴으로 구현하였는데, 우선 테스트할 때 어려움을 겪었습니다.

그렇다면 설계에 문제가 있었을 가능성이 높아보입니다.
좋은 객체 지향 설계를 판단하는 척도 중에 하나가 테스트 용이성 입니다.
테스트가 용이하지 않았다면 설계에 대해서 한번 의심해볼 상황인 것이죠.

> Getter와 Setter를 최대한 지양하기 위해 Lotto와 WinningLotto 사이에서 양방향 메서드를 사용하거나 DTO를 도입하였습니다. 구현하고보니 getter는 사용하지 않게 되었지만 양방향 메서드를 통한 Lotto와 WinningLotto가 서로를 참조하게 되는 상황입니다. 이에 대해서 그냥 getter를 일부 허용하는 것이 더 나은 방법일지 혹은 다른 개선 방안이 있을 지 궁금합니다.

getter 를 사용하지 말라는 이유에 대해서 한번 고민해보시면 좋을 것 같습니다.
왜 DTO를 쓰는게 getter를 사용하는 것보다 낫다고 생각하시는 공유해주시면 한번 이야기 해보시죠ㅎㅎ

Minuring replied:
> 그렇다면 설계에 문제가 있었을 가능성이 높아보입니다.
좋은 객체 지향 설계를 판단하는 척도 중에 하나가 테스트 용이성 입니다. 테스트가 용이하지 않았다면 설계에 대해서 한번 의심해볼 상황인 것이죠.

싱글톤 패턴을 사용하지 않았다면 테스트가 용이했을 것 같아요.
"유일하게 존재하는 당첨 로또" 라는 의미를 드러낼 방법을 생각해 보았는데 생성자를 막아야 의미가 있다고 생각하고, 다른 방법이 떠오르지 않습니다. 좋은 방법이 있을까요?

> getter를 사용하지 말라는 이유

객체간 메세지를 주고받으면서 협력하도록 하기 위함이라고 생각합니다. getter를 통해 객체의 캡슐화를 깨고 외부에서 로직을 수행하면 객체의 자율성이 없어지기 때문입니다. 즉 객체지향적이지 않게 됩니다.

> 왜 DTO를 쓰는 게 getter를 사용하는 것보다 낫다고 생각하는 지

도메인 객체의 멤버 변수에 대한 참조를 뷰와 가까운 쪽에서 getter를 통해 얻을 수 있게 되면 도메인 객체에 대한 변경 가능성이라는 위험이 생긴다고 생각했습니다.
그래서 데이터를 복사한 DTO를 사용해 뷰와 전달받고자 했는데, 현재 코드상으로는 얕은 복사를 하고있기 때문에 의미가 없다고 생각되네요.

그럼 아서는 도메인 객체를 불변으로 설계하고 필요시 출력을 위한 getter를 사용하되, 방어적 복사를 하도록 하면 DTO가 없어도 된다고 생각하시나요!?

Hyunta replied:
당첨 로또가 유일해야 할 정도로 중요하다면 클래스를 별도로 만드는게 낫지 않을까요? (Lotto, WinningLotto)
싱글턴 패턴은 객체에 특별한 의미를 드러내기 위해서 사용하는 것이 아니라, 보통 메모리 절약의 목적으로 사용됩니다.

> 객체간 메세지를 주고받으면서 협력하도록 하기 위함이라고 생각합니다. getter를 통해 객체의 캡슐화를 깨고 외부에서 로직을 수행하면 객체의 자율성이 없어지기 때문입니다. 즉 객체지향적이지 않게 됩니다.

그렇다면
Order 객체가 User 객체의 age가 필요해서 User.getAge() 를 호출하면 이것은 객체지향적이지 않은 것인가요?
이것도 메세지를 주고받아서 협력하는 관계로 볼 수 있지 않을까요?

> 그럼 아서는 도메인 객체를 불변으로 설계하고 필요시 출력을 위한 getter를 사용하되, 방어적 복사를 하도록 하면 DTO가 없어도 된다고 생각하시나요!?

네 도메인간 정보를 주고받는 내용은 방어적 복사로 충분하다고 생각합니다.

Hyunta reviewed (changes requested):
안녕하세요 포포,
우테코 7기에 합류하게 되신걸 축하드립니다! 🥳
첫 미션 리뷰어로 만나게 된 아서라고 합니다.

온보딩 미션 1단계는 미션 사이클을 경험하시는 것이 주된 목표이기 때문에 코드에 대한 리뷰보다는 전반적인 컨벤션, 커밋 로그, 문서 관리, 테스트와 같은 부분을 중점으로 리뷰를 진행했습니다.
2단계 진행하면서 코드에 대한 부분들도 리뷰를 해드릴테니 너무 걱정하지는 마세요.

테스트 관련해서 여러 시도를 많이 해주셨네요ㅎㅎ
몇가지 커멘트를 남겼는데 확인 부탁드릴게요
언제든지 궁금한 점이 있으시면 DM 주시거나 커멘트로 알려주세요!

Minuring committed:
refactor: 당첨로또 싱글톤 패턴 구현 철회

"유일하게 존재하는 당첨 로또" 라는 의미를 위해 구현했던 싱글톤 패턴으로 인해 테스트가 불가능함에 따라 싱글톤 패턴 구현을 철회합니다.

Minuring committed:
refactor: 도메인 객체들의 Factory 제거

별도의 생성 로직이 없고 단순히 도메인 객체를 생성하는 팩토리를 사용하는것은 오버 엔지니어링이라고 판단하여 제거합니다.

Minuring committed:
style: 파일 끝에 개행 추가

POSIX 표준에 따라 파일 끝에 개행을 추가합니다.

Minuring committed:
refactor(dto): DTO가 도메인 객체를 변경할 수 없도록 수정

기존 도메인 -> DTO 변환로직은 원본 객체에 대한 참조를 그대로 넘기기 때문에 안전하지 않다고 생각해 수정합니다.

- Lotto <-> LottoDto, Lottos <-> LottosDto
새로운 리스트를 만들어 반환하도록 변경

- Statistics <-> StatisticsDto
새로운 맵을 만들어 반환하도록 변경
(Key에 해당하는 PrizeTier는 enum이므로 불변하기 때문에 안전합니다.)

Minuring committed:
refactor(service): 단순 위임만 하는 서비스 제거

Minuring committed:
refactor: 가독성 향상을 위해 스트림 활용

반복문 대신 스트림 API를 이용했을 때 가독성이 향상된다고 생각한 코드들을 변경합니다.

Minuring committed:
style: 코드 컨벤션을 준수하도록 수정

Minuring committed:
refactor(model): 로또의 당첨 등수를 얻을 수 있게 변경

PrizeTier는 enum 상수이므로 안전하고 상금 금액 뿐만 아니라 다른 정보에도 편하게 접근할 수 있도록 하기 위해서
로또에 대해 상금만을 꺼낼 수 있었던 기존 방식 대신 PrizeTier를 꺼낼 수 있도록 변경합니다.

Minuring committed:
test(model): 로또 등수 판단 테스트

Hyunta reviewed (approved):
안녕하세요 포포,
고민을 많이 하면서 진행중이신게 느껴지네요.
2단계 리팩터링 진행하시면서 더 이야기 나눠보시죠.
